core/trino-main/src/main/java/io/trino/execution/TaskId.java

```diff
@@ -44,11 +44,12 @@ public static TaskId valueOf(String fullId)
         requireNonNull(stageId, "stageId is null");
         checkArgument(partitionId >= 0, "partitionId is negative: %s", partitionId);
         checkArgument(attemptId >= 0, "attemptId is negative: %s", attemptId);
+        checkArgument(fullId.equals(constructFullId(stageId, partitionId, attemptId)), "fullId does not match provided components");
```

Comment:
The problem is that this is now a runtime check. Before, it was physically impossible to construct an invalid TaskId object. Either the fullId should be provided during construction or the individual parts, but not both.

Reply:
https://github.com/openjdk/jdk/blob/dfd383224dbc2e41c9f44b1acd09ffb179cc38f3/src/hotspot/share/ci/ciField.cpp#L237 record classes are treated as effectively final (which isn't the case for ordinary final class with final fields). This allows a bunch of optimizations on top of that.